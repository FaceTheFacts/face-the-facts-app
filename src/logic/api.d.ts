export interface ApiSearchPolitician {
  id: number;
  label: string;
  party: ApiParty;
  image_url: string;
}

export interface ApiEntities<T> {
  items: Array<T>;
}

export interface ApiResponse<T> {
  items: T[];
}

export interface ApiEntity {
  politician: ApiEntities<ApiPolitician>;
}

export interface ApiPolitician {
  profile: ApiPoliticianProfile;
  positions: ApiPositions;
}

export interface IPoliticianContext {
  profile?: ApiPoliticianProfile;
  positions?: ApiPositions;
  speeches?: ApiPaginatedData<ApiSpeech>;
  news?: ApiNews;
  constituency?: ApiSearchPolitician[];
}

export interface ApiVoteAndPoll {
  Vote: ApiVote;
  Poll: ApiPoll;
}

export interface ApiVoteAndPoll {
  Vote: ApiVote;
  Poll: ApiPoll;
}

export interface ApiPoliticianProfile {
  id: number;
  entity_type: string;
  label: string;
  first_name: string;
  last_name: string;
  sex: string;
  year_of_birth: string;
  party: ApiParty;
  deceased: boolean;
  deceased_date: string;
  education: string;
  residence: string;
  occupations: string[];
  statistic_questions: string;
  statistic_questions_answered: string;
  qid_wikidata: string;
  field_title: string;
  sidejobs: ApiSidejob[];
  cvs: {
    polician_id: number;
    short_description: string;
    raw_text: string;
  }[];
  abgeordnetenwatch_url: string;
  weblinks: {
    id: number;
    politician_id: number;
    link: string;
  }[];
  votes_and_polls: ApiVoteAndPoll[];
  topic_ids_of_latest_committee: number[];
}

export interface ApiParty {
  id: number;
  label: string;
  party_style: ApiPartyStyle;
}

export interface ApiPartyStyle {
  id: number;
  display_name: string;
  foreground_color: string;
  background_color: string;
  border_color?: string;
}

export interface ApiVote {
  id: number;
  entity_type: string;
  label: string;
  api_url: string;
  mandate_id: number;
  fraction_id: number;
  poll_id: number;
  vote: Vote;
  reason_no_show: string;
  reason_no_show_other: string;
}

export interface ApiPoll {
  id: number;
  label: string;
  field_intro: string;
  field_poll_date: string;
  poll_passed: boolean;
}

export interface ApiSidejob {
  id: number;
  entity_type: string;
  label: string;
  income_level: string;
  interval: string;
  created: string;
  sidejob_organization: ApiSidejobOrganisation;
}

export interface ApiSidejobOrganisation {
  id: number;
  entity_type: string;
  label: string;
}

export interface ApiPositions {
  id: number;
  positions: ApiPosition[];
}

export interface ApiPosition {
  id: number;
  position: PositionAnswer;
  reason: string;
  position_statement: {
    statement: string;
  };
}

export interface ApiPollDetails {
  items: ApiEntities<ApiPollDetail>;
}

export interface ApiPollDetail {
  id: number;
  poll_id: number;
  fraction: {
    id: number;
    full_name: string;
    short_name: string;
    label: string;
  };
  total_yes: number;
  total_no: number;
  total_abstain: number;
  total_no_show: number;
}

interface SpeechAttribute {
  originID?: string;
  originMediaID?: string;
  creator: string;
  videoFileURI: string;
  dateEnd: string;
}

interface SpeechRelationship {
  agendaItem: {
    data: {
      attributes: {
        title: string;
      };
    };
  };
  people: {
    data: {
      attributes: {
        additionalInformation: {
          abgeordnetenwatchID: string;
        };
      };
    }[];
  };
}

interface ApiSpeechData {
  type: string;
  id: string;
  attributes: SpeechAttribute;
  relationships: SpeechRelationship;
}

interface SpeechResponse {
  meta: {
    api: unknown;
    requestStatus: string;
    results: {
      count: number;
      total: number;
    };
  };
  data: ApiSpeechData[];
}

export interface ApiPaginatedData<T> {
  items: T[];
  total: number;
  page: number;
  size: number;
  is_last_page: boolean;
}

export interface ApiSpeech {
  videoFileURI: string;
  title: string;
  date: string;
}

export interface ApiNews {
  items: ApiNewsArticle[];
}

export interface ApiNewsArticle {
  id: string;
  highlight: string;
  images: PoliTrackImage[];
  published: string;
  source: string;
  title: string;
  url: string;
}

export interface PoliTrackImage {
  url: string;
  title: string;
  height: number;
  width: number;
}

export type TopicIcon = {
  label: string;
  icon: string;
};

export type Vote = 'yes' | 'no' | 'abstain' | 'no_show';

export type PollResult = 'yes' | 'no';

export type PositionAnswer = 'agree' | 'disagree' | 'neutral';
