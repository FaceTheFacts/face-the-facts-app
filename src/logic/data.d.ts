export interface Dataset {
  politicians: Politician[];
  parties: Party[];
  polls: Poll[];
  pollTopics: PollTopic[];
  positions: Position[];
  constituencies: Constituency[];
  elections: Election[];
  committees: Committee[];
}

export interface Party {
  id: string;
  displayName: string;
  foregroundColor: string;
  backgroundColor: string;
  borderColor?: string;
}

export interface Committee {
  id: string;
  name: string;
  icon?: string;
}

export interface Poll {
  id: string;
  title: string;
  description: string;

  /**
   * A list of the IDs of all parties which participated in the poll
   */
  participatedParties: string[];

  /**
   * The number of votes per party. Every party has a dataset of 4 numbers. The parties
   * are concatenated behind each other in the same order as the `participatedParties`
   * field.
   *
   * A dataset of a party contains the number of votes in the order: yes, no, abstain, none
   *
   * Example:
   * participatedParties: [CDU, SPD]
   * votes: [0, 230, 4, 14, 2, 140, 2, 10]
   *
   * This means that the vote was as the following:
   *         | CDU | SDP
   * yes     |  0  |  2
   * no      | 230 | 140
   * abstain |  4  |  2
   * none    | 14  | 10
   */
  votes: number[];
  topic?: string[];
  date: string;
}

export interface PollTopic {
  id: string;
  title: string;
  icon: string;
}

export interface PollLink {
  link: string;
}

export interface Position {
  id: string;
  title: string;
  content: string;
}

export interface Constituency {
  id: string;
  name: string;
  number: number;
  zipCodes?: string[];
  locations?: string[];
  politicians: string[];
}

export interface Election {
  id: string;
  displayName: string;
  politicians: string[];
}

export interface Politician {
  id: string;
  popularity: number;
  name: string;
  partyId: string;
  constituency?: string;
  occupation?: string[];
  committees?: string[];
  votes?: Record<string, Vote>;
  sideJobs?: SideJob[];
  cv?: CVEntry[];
  links?: string[];
  positions?: PoliticianPosition[];
}

export type Vote = 'yes' | 'no' | 'abstain' | 'none';

export type PollResult = 'yes' | 'no';

export type PositionAnswer = 'yes' | 'no' | 'abstain';

export interface PoliticianPosition {
  id: string;
  reason: string;
  answer: PositionAnswer;
}

export type SideJobIncomeLevel =
  | '1'
  | '2'
  | '3'
  | '4'
  | '5'
  | '6'
  | '7'
  | '8'
  | '9'
  | '10';

export interface SideJob {
  job: string;
  organization: string;
  incomeLevel?: SideJobIncomeLevel;
  date?: string;
}

export interface CVEntry {
  date?: string;
  title: string;
}
