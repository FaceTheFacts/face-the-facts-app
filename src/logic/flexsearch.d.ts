declare module 'flexsearch' {
  export class Index {
    constructor(options?: any);

    add(id: number, token: string): this;

    append(id: number, token: string): this;

    search(query: string, limit?: number): number[];
  }

  export interface DocumentSearchResult {
    field: string;
    result: number[];
  }

  export class Document {
    constructor(options?: any);

    add(id: number, content: object): this;

    append(id: number, content: object): this;

    search(query: string, limit?: number): DocumentSearchResult[];

    searchAsync(query: string, limit?: number): Promise<DocumentSearchResult[]>;
  }
}
