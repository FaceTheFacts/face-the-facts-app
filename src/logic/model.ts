import {TrackedTextFeature} from 'react-native-camera';
import {PoliticianScanner} from './scanner';
import {makeToken} from './tokenizer';
import {createContext} from 'react';

export interface Politician {
  id: string;
  displayName: string;
  picture: string | null;
  party: string;
  popularity: number;
}

export interface Party {
  id: string;
  displayName: string;
  foregroundColor: string;
  backgroundColor: string;
  borderColor?: string;
}

export interface Election {
  id: string;
  displayName: string;
  politicians: string[];
}

export interface FaceTheFactsDataset {
  politicians: Politician[];
  parties: Party[];
  elections: Election[];
}

export class FaceTheFactsData {
  private readonly politicians: Map<string, Politician>;
  private readonly parties: Map<string, Party>;
  public readonly elections: Election[];

  private readonly scanTokens: Map<string, string[][]>;
  private readonly scanTokenMap: Map<string, Set<string>>;

  public constructor(dataset: FaceTheFactsDataset) {
    this.politicians = new Map(
      dataset.politicians.map(value => [value.id, value]),
    );
    this.parties = new Map(dataset.parties.map(value => [value.id, value]));
    this.elections = dataset.elections;

    this.scanTokens = new Map();
    this.scanTokenMap = new Map();
    dataset.politicians.forEach(({id, displayName}) => {
      const boundarySplit = displayName
        .split(/\b/)
        .filter(value => value.match(/\b/));
      const spaceSplit = displayName.split(/\s+/);
      const scanTokens: string[][] = [boundarySplit.map(makeToken)];
      if (spaceSplit.length !== boundarySplit.length) {
        scanTokens.push(spaceSplit.map(makeToken));
      }
      this.scanTokens.set(id, scanTokens);

      scanTokens.forEach(tokenBundle =>
        tokenBundle.forEach(token => {
          if (!this.scanTokenMap.get(token)?.add(id)) {
            this.scanTokenMap.set(token, new Set([id]));
          }
        }),
      );
    });
  }

  public lookupPolitician(id: string): Politician | null {
    return this.politicians.get(id) ?? null;
  }

  public lookupParty(id: string): Party | null {
    return this.parties.get(id) ?? null;
  }

  public scanPolitician(features: TrackedTextFeature[]): Politician | null {
    return new PoliticianScanner(this).analyze(features);
  }

  public lookupScanToken(token: string): Set<string> | null {
    return this.scanTokenMap.get(token) ?? null;
  }

  public lookupScanTokens(politicianId: string): string[][] | null {
    return this.scanTokens.get(politicianId) ?? null;
  }
}

export const DataContext = createContext<FaceTheFactsData>(null as any);
