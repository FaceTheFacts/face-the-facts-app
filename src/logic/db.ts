import SQLite, {SQLiteDatabase} from 'react-native-sqlite-storage';

SQLite.enablePromise(true);

export interface HistoryItem {
  politicianId: number;
  date: Date;
}

export class DBManager {
  private followedIds: Set<number> | null = null;
  private historyItems: HistoryItem[] | null = null;
  private database: SQLiteDatabase | null = null;

  public constructor(private readonly politicians: Map<string, string>) {}

  public async getFollowedIds(): Promise<Set<number>> {
    await this.loadFollowedIds();
    return this.followedIds!;
  }

  public async getHistoryItems(): Promise<HistoryItem[]> {
    await this.loadHistoryItems();
    return this.historyItems!;
  }

  public async pushHistoryItem(politicianId: number): Promise<void> {
    const date = new Date();
    await this.openDatabase();
    await this.database!.executeSql(
      'DELETE FROM history WHERE politician_id = ?',
      [politicianId.toString()],
    );
    await this.database!.executeSql(
      'INSERT OR IGNORE INTO history(politician_id, date) VALUES (?, ?)',
      [politicianId.toString(), date.toISOString()],
    );
    if (this.historyItems) {
      for (let i = 0; i < this.historyItems.length; i++) {
        if (this.historyItems[i].politicianId === politicianId) {
          this.historyItems.splice(i, 1);
          break;
        }
      }
      this.historyItems.push({
        politicianId,
        date,
      });
    }
  }

  public async isIdFollowed(politicianId: number): Promise<boolean> {
    await this.loadFollowedIds();
    if (this.followedIds!.has(politicianId)) {
      return true;
    }
    return false;
  }

  public async pushFollowId(politicianId: number): Promise<void> {
    const date = new Date();
    await this.openDatabase();
    await this.database!.executeSql(
      'INSERT OR IGNORE INTO follow(id, date) VALUES (?, ?)',
      [politicianId, date.toISOString()],
    );
    this.followedIds!.add(politicianId);
  }

  public async removeFollowedId(politicianId: number): Promise<void> {
    await this.openDatabase();
    await this.database!.executeSql('DELETE FROM follow WHERE id = ?', [
      politicianId,
    ]);
    this.followedIds!.delete(politicianId);
  }

  private async loadHistoryItems(): Promise<void> {
    if (this.historyItems) {
      return;
    }

    await this.openDatabase();
    const [{rows}] = await this.database!.executeSql(
      'SELECT id, politician_id, date FROM history ORDER BY date',
    );
    await Promise.all(
      rows
        .raw()
        .filter(({politician_id}) => !this.politicians.has(politician_id))
        .map(({id}) =>
          this.database!.executeSql('DELETE FROM history WHERE id = ?', [id]),
        ),
    );
    this.historyItems = rows
      .raw()
      .filter(({politician_id}) => this.politicians.has(politician_id))
      .map(({politician_id, date}) => ({
        politicianId: politician_id,
        date: new Date(date),
      }));
  }

  private async loadFollowedIds(): Promise<void> {
    if (this.followedIds) {
      return;
    }

    await this.openDatabase();
    const [{rows}] = await this.database!.executeSql(
      'SELECT id FROM follow ORDER BY date',
    );
    this.followedIds = new Set(rows.raw().map(({id}) => id));
  }

  private async openDatabase(): Promise<void> {
    if (this.database) {
      return;
    }

    this.database = await SQLite.openDatabase({
      name: 'database',
      location: 'Documents',
    });
    await this.database.executeSql(`CREATE TABLE IF NOT EXISTS history (
        id            INTEGER   NOT NULL PRIMARY KEY AUTOINCREMENT,
        politician_id TEXT      NOT NULL UNIQUE,
        date          TIMESTAMP NOT NULL
    )`);
    await this.database.executeSql(`CREATE TABLE IF NOT EXISTS follow (
        id            INTEGER   NOT NULL PRIMARY KEY,
        date          TIMESTAMP NOT NULL
    )`);
    await this.database.executeSql(
      "DELETE FROM history WHERE date < DATETIME('now', '-1 month')",
    );
  }
}
