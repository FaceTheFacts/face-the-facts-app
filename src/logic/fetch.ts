export async function fetch_data(url: string) {
    const BASE_URL = 'https://api.facethefacts-api.de/';
    try {
	const response = await fetch(BASE_URL + url);
    const json = await response.json()
    console.log(url)
	return json;
    } catch(error) {
        console.error(error);
    }
}