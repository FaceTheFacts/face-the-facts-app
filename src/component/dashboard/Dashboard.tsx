import {NavigationContext} from '@react-navigation/native';
import React, {useContext, useEffect, useState} from 'react';
import {
  View,
  Text,
  TouchableOpacity,
  StyleSheet,
  ScrollView,
  useWindowDimensions,
} from 'react-native';
import {useQuery} from 'react-query';
import {
  ApiPaginatedData,
  ApiSpeechBundestag,
  ApiSidejobsBundestag,
  ApiPollBundestagData,
  ApiBundestagPartyDonation,
} from '../../logic/api';
import {fetch_api} from '../../logic/fetch';
import {DataContext} from '../../logic/model';
import {Colors} from '../../theme';
import {formatDate} from '../../utils/util';
import DashboardPollCard from '../poll/DashboardPollCard';
import SkeletonDashboard from '../skeleton/SkeletonDashboard';
import SpeechCard from '../speech/SpeechCard';
import SideJobCard from '../SideJobCard';
import ErrorCard from '../Error';
import PartyDonationCard from '../partydonation/PartydonationCard';

const Dashboard = () => {
  const navigator = useContext(NavigationContext);
  const database = useContext(DataContext);
  const [followedIds, setFollowedIds] = useState<Set<number>>(new Set());
  useEffect(() => {
    database.dbManager.getFollowedIds().then(setFollowedIds);
  }, [database]);
  let query = '';
  Array.from(followedIds).map(id => (query += `follow_ids=${id}&`));

  const {
    data: speeches,
    isLoading: speechesLoading,
    isError: speechesError,
  } = useQuery<ApiPaginatedData<ApiSpeechBundestag> | undefined, Error>(
    'speeches: Bundestag page:1',
    () =>
      fetch_api<ApiPaginatedData<ApiSpeechBundestag>>(
        'bundestag/speeches?page=1',
      ),
    {
      staleTime: 60 * 10000000, // 10000 minute = around 1 week
      cacheTime: 60 * 10000000,
    },
  );

  const {
    data: polls,
    isLoading: pollsLoading,
    isError: pollsError,
  } = useQuery<ApiPollBundestagData[] | undefined, Error>(
    `polls: Bundestag ${query}`,
    () => fetch_api<ApiPollBundestagData[]>(`bundestag/polls?${query}`),
    {
      staleTime: 60 * 10000000, // 10000 minute = around 1 week
      cacheTime: 60 * 10000000,
    },
  );

  const {
    data: sidejobs,
    isLoading: sidejobsLoading,
    isError: sidejobsError,
  } = useQuery<ApiSidejobsBundestag[] | undefined, Error>(
    'sidejobs: Bundestag',
    () => fetch_api<ApiSidejobsBundestag[]>('bundestag/sidejobs?size=5'),
    {
      staleTime: 60 * 10000000, // 10000 minute = around 1 week
      cacheTime: 60 * 10000000,
    },
  );

  const {
    data: partydonations,
    isLoading: partydonationsLoading,
    isError: partydonationsError,
  } = useQuery<ApiBundestagPartyDonation[] | undefined, Error>(
    'partdonations: Bundestag',
    () => fetch_api<ApiBundestagPartyDonation[]>('homepagepartydonations'),
    {
      staleTime: 60 * 10000000, // 10000 minute = around 1 week
      cacheTime: 60 * 10000000,
    },
  );
  const {width} = useWindowDimensions();
  if (
    pollsLoading ||
    speechesLoading ||
    sidejobsLoading ||
    partydonationsLoading
  ) {
    return <SkeletonDashboard />;
  }
  if (pollsError || sidejobsError || speechesError || partydonationsError) {
    return <ErrorCard />;
  }

  return (
    <ScrollView showsVerticalScrollIndicator={false}>
      {partydonations && partydonations.length > 0 && (
        <>
          <View style={styles.header}>
            <Text style={styles.headerText}>Parteispenden</Text>
            {/* <TouchableOpacity
              style={styles.moreBtn}
              onPress={() => {
                navigator?.navigate('DashboardPartyDonations', {polls});
              }}>
              <Text style={styles.btnText}>mehr</Text>
            </TouchableOpacity> */}
          </View>
          <ScrollView
            horizontal
            showsHorizontalScrollIndicator={false}
            style={styles.scrollContainer}>
            {partydonations.map((partydonation, index) => (
              <PartyDonationCard
                key={index}
                party={partydonation.party}
                donations={partydonation.donations_over_96_months.reverse()}
                donations_total={partydonation.donations_total}
              />
            ))}
          </ScrollView>
        </>
      )}
      {speeches && (
        <View style={styles.topContainer}>
          <View style={styles.header}>
            <Text style={styles.headerText}>Reden</Text>
            <TouchableOpacity
              style={styles.moreBtn}
              onPress={() => {
                navigator?.navigate('DashboardSpeeches', {
                  speeches: speeches,
                });
              }}>
              <Text style={styles.btnText}>mehr</Text>
            </TouchableOpacity>
          </View>

          <ScrollView
            horizontal
            contentContainerStyle={{paddingRight: 24}}
            showsHorizontalScrollIndicator={false}
            style={styles.scrollContainer}>
            {speeches.items.map((speech, index) => (
              <SpeechCard
                key={index}
                politicianId={speech.speaker.id}
                politicianName={speech.speaker.label}
                party={speech.speaker.party}
                title={speech.title}
                date={formatDate(speech.date)}
                video={speech.videoFileURI}
                cardHeight={143}
                cardWidth={width * 0.71}
                verticalScroll
              />
            ))}
          </ScrollView>
        </View>
      )}

      {polls && (
        <View style={styles.container}>
          <View style={styles.header}>
            <Text style={styles.headerText}>Abstimmungen</Text>
            <TouchableOpacity
              style={styles.moreBtn}
              onPress={() => {
                navigator?.navigate('DashboardPolls', {polls});
              }}>
              <Text style={styles.btnText}>mehr</Text>
            </TouchableOpacity>
          </View>
          <View style={styles.scrollContainer}>
            {polls.map((poll, index) => (
              <DashboardPollCard
                key={index}
                poll={poll.poll}
                result={poll.result}
                politicians={poll.politicians}
                last_politician={poll.last_politician}
              />
            ))}
          </View>
        </View>
      )}
      {sidejobs && (
        <View style={styles.container}>
          <View style={styles.header}>
            <Text style={styles.headerText}>Nebent√§tigkeiten</Text>
            <TouchableOpacity
              style={styles.moreBtn}
              onPress={() => {
                navigator?.navigate('DashboardSidejobs');
              }}>
              <Text style={styles.btnText}>mehr</Text>
            </TouchableOpacity>
          </View>
          <ScrollView
            horizontal
            contentContainerStyle={{paddingRight: 24}}
            showsHorizontalScrollIndicator={false}
            style={styles.scrollContainer}>
            {sidejobs.map((sidejob, index) => (
              <TouchableOpacity
                key={index}
                onPress={() => {
                  database.dbManager.pushHistoryItem(sidejob.politician.id);
                  navigator?.navigate('Politician', {
                    politicianId: sidejob.politician.id,
                    politicianName: sidejob.politician.label,
                    party: sidejob.politician.party,
                    toSideJobs: true,
                  });
                }}>
                <SideJobCard
                  politicianId={sidejob.politician.id}
                  politicianName={sidejob.politician.label}
                  party={sidejob.politician.party}
                  label={sidejob.sidejob.label}
                  date={sidejob.sidejob.created}
                  income={sidejob.sidejob.income_level}
                  organization={sidejob.sidejob.sidejob_organization.label}
                />
              </TouchableOpacity>
            ))}
          </ScrollView>
        </View>
      )}
    </ScrollView>
  );
};

const styles = StyleSheet.create({
  topContainer: {
    flex: 1,
    justifyContent: 'flex-start',
    marginTop: 4,
    marginBottom: 12,
  },
  container: {
    flex: 1,
    justifyContent: 'flex-start',
    marginVertical: 12,
  },
  header: {
    flexDirection: 'row',
    paddingHorizontal: 16,
    justifyContent: 'space-between',
    alignItems: 'center',
    marginVertical: 8,
  },
  headerText: {
    fontSize: 17,
    fontWeight: '600',
    color: Colors.baseWhite,
  },
  btnText: {
    fontSize: 13,
    fontWeight: '600',
    color: Colors.baseWhite,
  },
  moreBtn: {
    borderColor: Colors.white40,
    borderWidth: 2,
    paddingVertical: 6,
    paddingHorizontal: 10,
    borderRadius: 4,
  },
  scrollContainer: {
    paddingHorizontal: 16,
  },
});

export default Dashboard;
