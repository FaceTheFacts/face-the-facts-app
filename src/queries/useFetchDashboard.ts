import {useState} from 'react';
import {Alert} from 'react-native';
import {fetch_api} from '../logic/fetch';
import {
  ApiPaginatedData,
  ApiSpeechBundestag,
  ApiSidejobsBundestag,
  ApiPollBundestagData,
  ApiBundestagPartyDonation,
} from '../logic/api';
import {useQueries} from 'react-query';

const queryFunctions = {
  speeches: () =>
    fetch_api<ApiPaginatedData<ApiSpeechBundestag>>(
      'bundestag/speeches?page=1',
    ),
  polls: (pollquery: string) =>
    fetch_api<ApiPollBundestagData[]>(`bundestag/polls?${pollquery}`),
  sidejobs: () =>
    fetch_api<ApiSidejobsBundestag[]>('bundestag/sidejobs?size=5'),
  donations: () =>
    fetch_api<ApiBundestagPartyDonation[]>('homepagepartydonations'),
};

export const useFetchDashboard = (followIds: Set<number>) => {
  const [lastAlertTime, setLastAlertTime] = useState<number>(0);

  const handleError = (error: any) => {
    const errorMessage = error ? error.toString() : '';
    const currentTime = Date.now();

    if (
      errorMessage.includes('Das Zertifikat für diesen Server ist ungültig') &&
      currentTime - lastAlertTime > 5000
    ) {
      setLastAlertTime(currentTime);
      Alert.alert(
        'Sicherheitswarnung',
        'Wir haben ein mögliches Sicherheitsproblem festgestellt. Bitte stellen Sie sicher, dass Sie mit einem sicheren Netzwerk verbunden sind, und versuchen Sie es erneut.',
        [{text: 'OK', onPress: () => console.log('OK gedrückt')}],
        {cancelable: false},
      );
    }
  };

  const pollquery = Array.from(followIds)
    .map(id => `follow_ids=${id}&`)
    .join('');
  return useQueries([
    {
      queryKey: 'donations: Bundestag',
      queryFn: queryFunctions.donations,
      staleTime: 60 * 10000000, // 10000 minute = around 1 week
      cacheTime: 60 * 10000000,
    },
    {
      queryKey: 'speeches: Bundestag page:1',
      queryFn: queryFunctions.speeches,
      staleTime: 60 * 10000000, // 10000 minute = around 1 week
      cacheTime: 60 * 10000000,
    },
    {
      queryKey: `polls: Bundestag ${pollquery}`,
      queryFn: () => queryFunctions.polls(pollquery),
      staleTime: 60 * 10000000, // 10000 minute = around 1 week
      cacheTime: 60 * 10000000,
      onError: handleError,
    },
    {
      queryKey: 'sidejobs: Bundestag',
      queryFn: queryFunctions.sidejobs,
      staleTime: 60 * 10000000, // 10000 minute = around 1 week
      cacheTime: 60 * 10000000,
      onError: handleError,
    },
  ]);
};
