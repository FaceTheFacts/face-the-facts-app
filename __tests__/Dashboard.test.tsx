import React from 'react';
//import {render} from '@testing-library/react-native';
import Dashboard from '../src/component/dashboard/Dashboard';
import {useFetchDashboard} from '../src/queries/useFetchDashboard';
import {DataContext} from '../src/logic/model';
import renderer from 'react-test-renderer';

jest.mock('react-native-chart-kit', () => ({
  LineChart: () => null,
}));

jest.mock('react-native-video', () => ({
  Video: () => null,
}));

jest.mock('../src/queries/useFetchDashboard', () => ({
  useFetchDashboard: jest.fn(),
}));

jest.spyOn(React, 'useContext').mockImplementation(() => {
  // return the mock object for the DataContext
  return DataContext;
});

beforeEach(() => {
  // mock the useFetchDashboard hook to return an array of mock query objects with the isLoading property set to true
  (useFetchDashboard as jest.Mock).mockImplementation(() => [
    {isLoading: true},
    {isLoading: true},
    {isLoading: true},
    {isLoading: true},
  ]);
});

it('renders correctly', () => {
  const tree = renderer.create(<Dashboard />).toJSON();
  expect(tree).toMatchSnapshot();
});
