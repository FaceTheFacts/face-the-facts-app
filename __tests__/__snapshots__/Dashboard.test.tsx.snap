// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "backgroundColor": "#2A2B36",
        "flex": 0,
      }
    }
  />,
  <View
    style={
      Object {
        "backgroundColor": "#181924",
      }
    }
  >
    <RNSVGSvgView
      bbHeight="100%"
      bbWidth="100%"
      focusable={false}
      interval={0.25}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {},
          Object {
            "flex": 0,
            "height": "100%",
            "width": "100%",
          },
        ]
      }
    >
      <RNSVGGroup>
        <RNSVGRect
          clipPath="l0vfr7l-animated-diff"
          fill={
            Array [
              1,
              "l0vfr7l-diff",
            ]
          }
          height="100%"
          propList={
            Array [
              "fill",
            ]
          }
          width="100%"
          x="0"
          y="0"
        />
        <RNSVGDefs>
          <RNSVGClipPath
            name="l0vfr7l-animated-diff"
          >
            <RNSVGRect
              height="16"
              rx="4"
              ry="4"
              width="92"
              x="12"
              y="12"
            />
            <RNSVGRect
              height="120"
              rx="4"
              ry="4"
              width={726}
              x="12"
              y="40"
            />
            <RNSVGRect
              height="16"
              rx="4"
              ry="4"
              width="92"
              x="12"
              y="184"
            />
            <RNSVGRect
              height="80"
              rx="4"
              ry="4"
              width={726}
              x="12"
              y="212"
            />
            <RNSVGRect
              height="80"
              rx="4"
              ry="4"
              width={726}
              x="12"
              y="304"
            />
            <RNSVGRect
              height="80"
              rx="4"
              ry="4"
              width={726}
              x="12"
              y="400"
            />
          </RNSVGClipPath>
          <RNSVGLinearGradient
            gradient={
              Array [
                0,
                -14013642,
                0.5,
                -1118482,
                1,
                -14013642,
              ]
            }
            gradientTransform={null}
            gradientUnits={0}
            name="l0vfr7l-diff"
            x1="-100%"
            x2="0%"
            y1={0}
            y2={0}
          />
        </RNSVGDefs>
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>,
]
`;
